### PR TITLE
Fixes for SNMP issue and records to return value

### PR DESCRIPTION
**Pull request by abimishr**
Problem: If we are having null values in credentials (v2) it throws nonetype error. Also, we have issues in API if the start_index > 500.
Fix: Made null to empty list for resolution of the error related to credentials. For API, I have added one more case which dynamically checks the records_to_return. Additionally, have added documentation for the HTTP credentials dict.
Test cases: Muti range with records_to return value of 720 and empty SNMP lists for failure.

### CODE REVIEW

File: plugins/modules/discovery_intent.py [hunk 1]
```diff
@@ -70,9 +70,35 @@
       http_read_credential:
         description: HTTP read credentials for hosting a device
         type: dict
+        suboptions:
+            password:
+                description: HTTP(S) password and is mandatory for using HTTP credentials.
```
Comment (madhansansel) [inline plugins/modules/discovery_intent.py:75]: Can it be bit more detail?
            description: The password for HTTP(S) authentication. Mandatory for utilizing HTTP credentials.

File: plugins/modules/discovery_intent.py [hunk 2]
```diff
@@ -70,9 +70,35 @@
...
+                type: str
+            port:
+                description: HTTP(S) port and is mandatory for using HTTP credentials.
```
Comment (madhansansel) [inline plugins/modules/discovery_intent.py:78]: description: The HTTP(S) port number, which is mandatory for using HTTP credentials.
(or)
description: This is the HTTP(S) port. It is mandatory for the usage of HTTP credentials

File: plugins/modules/discovery_intent.py [hunk 3]
```diff
@@ -70,9 +70,35 @@
...
+                type: int
+            secure:
+                description: Flag for HTTP(S) and is not mandatory for using HTTP credentials.
```
Comment (madhansansel) [inline plugins/modules/discovery_intent.py:81]: description: A flag indicating whether HTTP(S) is secure. It is optional and not mandatory for using HTTP credentials.
(or)
description: This is a flag for HTTP(S). Its usage is not mandatory for HTTP credentials.

File: plugins/modules/discovery_intent.py [hunk 4]
```diff
@@ -70,9 +70,35 @@
...
+                type: bool
+            username:
+                description: HTTP(S) username and is mandatory for using HTTP credentials.
```
Comment (madhansansel) [inline plugins/modules/discovery_intent.py:84]: description: The username for HTTP(S) authentication, which is mandatory when using HTTP credentials.
or
username: description: This is the HTTP(S) username. It is mandatory for the usage of HTTP credentials

File: plugins/modules/discovery_intent.py [hunk 5]
```diff
@@ -70,9 +70,35 @@
       http_read_credential:
         description: HTTP read credentials for hosting a device
         type: dict
+        suboptions:
```
Comment (madhansansel) [inline plugins/modules/discovery_intent.py:73]: We can rearrange the suboptions. We can list all the mandatory in the beginning...
username. -- Mandatory
password -- Mandatory
port -- Mandatory
secure -- Optional

File: plugins/modules/discovery_intent.py [hunk 6]
```diff
@@ -70,9 +70,35 @@
...
+            port:
+                description: HTTP(S) port and is mandatory for using HTTP credentials.
+                type: int
+            secure:
+                description: Flag for HTTP(S) and is not mandatory for using HTTP credentials.
+                type: bool
+            username:
+                description: HTTP(S) username and is mandatory for using HTTP credentials.
+                type: str
       http_write_credential:
```
Comment (madhansansel) [inline plugins/modules/discovery_intent.py:86]: Same as like above

File: plugins/modules/discovery_intent.py [hunk 7]
```diff
@@ -548,6 +593,32 @@ def create_params(self, credential_ids=None, ip_address_list=None):
         if credential_ids is None:
             credential_ids = []
 
+        http_read_credential = self.validated_config[0].get('http_read_credential')
+        http_write_credential = self.validated_config[0].get('http_write_credential')
+        if http_read_credential:
+            if not (http_read_credential.get('password') and isinstance(http_read_credential.get('password'), str)):
+                msg = "Http Read Credential must have a password of type string"
```
Comment (madhansansel) [inline plugins/modules/discovery_intent.py:600]: for better clarity
msg = "The password for the HTTP read credential must be of string type."

File: plugins/modules/discovery_intent.py [hunk 8]
```diff
@@ -548,6 +593,32 @@ def create_params(self, credential_ids=None, ip_address_list=None):
...
+            if not (http_read_credential.get('username') and isinstance(http_read_credential.get('username'), str)):
+                msg = "Http Read Credential must have a username of type string"
```
Comment (madhansansel) [inline plugins/modules/discovery_intent.py:602]: msg = "The username for the HTTP read credential must be of string type.

File: plugins/modules/discovery_intent.py [hunk 9]
```diff
@@ -548,6 +593,32 @@ def create_params(self, credential_ids=None, ip_address_list=None):
...
+            if not (http_read_credential.get('password') and isinstance(http_read_credential.get('password'), int)):
+                msg = "Http Read Credential must have port of type integer"
```
Comment (madhansansel) [inline plugins/modules/discovery_intent.py:604]: msg = "The port for the HTTP Read Credential must be of integer type.

File: plugins/modules/discovery_intent.py [hunk 10]
```diff
@@ -548,6 +593,32 @@ def create_params(self, credential_ids=None, ip_address_list=None):
...
+            if not isinstance(http_read_credential.get('secure'), bool):
+                msg = "Secure must be of type bool"
+            self.log(msg, "CRITICAL")
+            self.module.fail_json(msg=msg)
```
Comment (madhansansel) [inline plugins/modules/discovery_intent.py:608]: I don't understand why do we fail the module if all are good?

File: plugins/modules/discovery_intent.py [hunk 11]
```diff
@@ -548,6 +593,32 @@ def create_params(self, credential_ids=None, ip_address_list=None):
...
+            if not (http_write_credential.get('password') and isinstance(http_read_credential.get('password'), str)):
+                msg = "Http Write Credential must have a password of type string"
+            if not (http_write_credential.get('username') and isinstance(http_read_credential.get('username'), str)):
+                msg = "Http Write Credential must have a username of type string"
+            if not (http_write_credential.get('password') and isinstance(http_read_credential.get('password'), int)):
+                msg = "Http Write Credential must have port of type integer"
+            if not isinstance(http_write_credential.get('secure'), bool):
+                msg = "Secure must be of type bool"
+            self.log(msg, "CRITICAL")
+            self.module.fail_json(msg=msg)
```
Comment (madhansansel) [inline plugins/modules/discovery_intent.py:620]: I don't understand why do we fail the module if all are good?

File: plugins/modules/discovery_intent.py [hunk 12]
```diff
@@ -548,6 +593,32 @@ def create_params(self, credential_ids=None, ip_address_list=None):
...
+            if not (http_read_credential.get('username') and isinstance(http_read_credential.get('username'), str)):
+                msg = "Http Read Credential must have a username of type string"
+            if not (http_read_credential.get('password') and isinstance(http_read_credential.get('password'), int)):
+                msg = "Http Read Credential must have port of type integer"
+            if not isinstance(http_read_credential.get('secure'), bool):
+                msg = "Secure must be of type bool"
+            self.log(msg, "CRITICAL")
+            self.module.fail_json(msg=msg)
+
+        if http_write_credential:
```
Comment (madhansansel) [inline plugins/modules/discovery_intent.py:610]: Same comments as above

**Review by madhansansel — CHANGES_REQUESTED** [review]

File: plugins/modules/discovery_intent.py (hunk 1 above)
Comment (abimishr) [reply plugins/modules/discovery_intent.py]: Addressed

File: plugins/modules/discovery_intent.py (hunk 2 above)
Comment (abimishr) [reply plugins/modules/discovery_intent.py]: Addressed

File: plugins/modules/discovery_intent.py (hunk 3 above)
Comment (abimishr) [reply plugins/modules/discovery_intent.py]: Addressed

File: plugins/modules/discovery_intent.py (hunk 4 above)
Comment (abimishr) [reply plugins/modules/discovery_intent.py]: Addressed

File: plugins/modules/discovery_intent.py (hunk 5 above)
Comment (abimishr) [reply plugins/modules/discovery_intent.py:73]: Addressed

File: plugins/modules/discovery_intent.py (hunk 6 above)
Comment (abimishr) [reply plugins/modules/discovery_intent.py:86]: Addressed

File: plugins/modules/discovery_intent.py (hunk 7 above)
Comment (abimishr) [reply plugins/modules/discovery_intent.py]: Addressed

File: plugins/modules/discovery_intent.py (hunk 8 above)
Comment (abimishr) [reply plugins/modules/discovery_intent.py]: Addressed

File: plugins/modules/discovery_intent.py (hunk 9 above)
Comment (abimishr) [reply plugins/modules/discovery_intent.py]: Addressed

File: plugins/modules/discovery_intent.py (hunk 10 above)
Comment (abimishr) [reply plugins/modules/discovery_intent.py]: Addressed

File: plugins/modules/discovery_intent.py (hunk 11 above)
Comment (abimishr) [reply plugins/modules/discovery_intent.py]: Thanks for bringing it up, have corrected it

File: plugins/modules/discovery_intent.py (hunk 12 above)
Comment (abimishr) [reply plugins/modules/discovery_intent.py:620]: corrected

**Review by madhansansel — APPROVED** [review]

**Pull request merged**